drivers/modem/modem_cellular.c

```diff
@@ -0,0 +1,1226 @@
+	int ret;
+
+	switch (action) {
+	case PM_DEVICE_ACTION_RESUME:
+		atomic_set(&data->suspend_atomic, 0);
+
+		modem_cellular_delegate_event(data, MODEM_CELLULAR_EVENT_RESUME);
+
+		ret = 0;
+
+		break;
+
+	case PM_DEVICE_ACTION_SUSPEND:
+		atomic_set(&data->suspend_atomic, 1);
+
+		ret = k_sem_take(&data->suspended_sem, K_SECONDS(30));
+
+		break;
+
+	default:
+		ret = -ENOTSUP;
+
+		break;
+	}
+
+	return ret;
+}
+#endif /* CONFIG_PM_DEVICE */
+
+static int modem_cellular_init(const struct device *dev)
+{
+	struct modem_cellular_data *data = (struct modem_cellular_data *)dev->data;
+	struct modem_cellular_config *config = (struct modem_cellular_config *)dev->config;
+
+	data->dev = dev;
+
+	k_work_init_delayable(&data->timeout_work, modem_cellular_timeout_handler);
+
+	k_work_init(&data->event_dispatch_work, modem_cellular_event_dispatch_handler);
+	ring_buf_init(&data->event_rb, sizeof(data->event_buf), data->event_buf);
+
+	atomic_set(&data->suspend_atomic, 0);
+	k_sem_init(&data->suspended_sem, 0, 1);
+
+	{
+		const struct modem_backend_uart_config uart_backend_config = {
+			.uart = config->uart,
+			.receive_buf = data->uart_backend_receive_buf,
+			.receive_buf_size = ARRAY_SIZE(data->uart_backend_receive_buf),
+			.transmit_buf = data->uart_backend_transmit_buf,
+			.transmit_buf_size = ARRAY_SIZE(data->uart_backend_transmit_buf),
+		};
+
+		data->uart_pipe = modem_backend_uart_init(&data->uart_backend,
+							  &uart_backend_config);
+	}
+
+	{
+		const struct modem_cmux_config cmux_config = {
+			.callback = modem_cellular_cmux_handler,
+			.user_data = data,
+			.receive_buf = data->cmux_receive_buf,
+			.receive_buf_size = ARRAY_SIZE(data->cmux_receive_buf),
+			.transmit_buf = data->cmux_transmit_buf,
+			.transmit_buf_size = ARRAY_SIZE(data->cmux_transmit_buf),
+		};
+
+		modem_cmux_init(&data->cmux, &cmux_config);
+	}
+
+	{
+		const struct modem_cmux_dlci_config dlci1_config = {
+			.dlci_address = 1,
+			.receive_buf = data->dlci1_receive_buf,
+			.receive_buf_size = ARRAY_SIZE(data->dlci1_receive_buf),
+		};
+
+		data->dlci1_pipe = modem_cmux_dlci_init(&data->cmux, &data->dlci1,
+							&dlci1_config);
+	}
+
+	{
+		const struct modem_cmux_dlci_config dlci2_config = {
+			.dlci_address = 2,
+			.receive_buf = data->dlci2_receive_buf,
+			.receive_buf_size = ARRAY_SIZE(data->dlci2_receive_buf),
+		};
+
+		data->dlci2_pipe = modem_cmux_dlci_init(&data->cmux, &data->dlci2,
+							&dlci2_config);
+	}
+
+	{
+		const struct modem_chat_config chat_config = {
+			.user_data = data,
+			.receive_buf = data->chat_receive_buf,
+			.receive_buf_size = ARRAY_SIZE(data->chat_receive_buf),
+			.delimiter = data->chat_delimiter,
+			.delimiter_size = ARRAY_SIZE(data->chat_delimiter),
+			.filter = data->chat_filter,
+			.filter_size = ARRAY_SIZE(data->chat_filter),
+			.argv = data->chat_argv,
+			.argv_size = ARRAY_SIZE(data->chat_argv),
+			.unsol_matches = NULL,
+			.unsol_matches_size = 0,
```

Comment:
Should we add a small set of unsol_matches. Might be nice for users starting. Maybe monitoring +CEREG in case the network drops.